AdminUi/src/AdminUi/ClientApp/src/app/services/breadcrumb-service/breadcrumb.service.spec.ts

```diff
@@ -0,0 +1,171 @@
+import { TestBed } from "@angular/core/testing";
+import { BreadcrumbService } from "./breadcrumb.service";
+import { ActivatedRoute, NavigationEnd, Router } from "@angular/router";
+import { BehaviorSubject } from "rxjs";
+
+class MockActivatedRoute {
+    private readonly subject = new BehaviorSubject(this.testParams);
+    public params = this.subject.asObservable();
+
+    private _testParams!: {};
+    public get testParams() {
+        return this._testParams;
+    }
+    public set testParams(params: {}) {
+        this._testParams = params;
+        this.subject.next(params);
+    }
+}
+
+describe("BreadcrumbService", function () {
+    let breadcrumbService: BreadcrumbService;
+    let mockRouter: Router;
+    let mockActivatedRoute: MockActivatedRoute;
+
+    const initialHistory = [
+        { label: "Home", url: "/home" },
+        { label: "Details", url: "/details" },
+        { label: "Products", url: "/products" }
+    ];
+
+    beforeEach(function () {
+        mockRouter = {
+            events: new BehaviorSubject(new NavigationEnd(0, "", ""))
+        } as any;
+
+        mockActivatedRoute = new MockActivatedRoute();
+
+        TestBed.configureTestingModule({
+            providers: [BreadcrumbService, { provide: Router, useValue: mockRouter }, { provide: ActivatedRoute, useValue: mockActivatedRoute }]
+        });
+        breadcrumbService = TestBed.inject(BreadcrumbService);
+    });
+
+    it("should be created", async function () {
+        await expect(breadcrumbService).toBeTruthy();
+    });
+
+    it("should return an empty array initially", async function () {
+        // Arrange
+
+        // Act
+        const result = breadcrumbService.getBreadcrumbHistory();
+
+        // Assert
+        await expect(result.length).toEqual(0);
+    });
+
+    it("should return the breadcrumb history array", async function () {
+        // Arrange
+        const expectedHistory = [
+            { label: "Home", url: "/home" },
+            { label: "Details", url: "/details" },
+            { label: "Products", url: "/products" }
+        ];
+
+        // Act
+        expectedHistory.forEach((trail) => {
```

Comment:
Changing from forEach to map function is duable in 2 out of 3 test cases, in test case "should not modify the internal breadcrumb history array when modifying the result" i was trying to create a deep copy of expectedHistory, so when i use map on something like that it is not duable because you can achieve a deep copy using map if your array contains only simple objects (no nested objects or arrays).

Reply:
I think you misunderstood me. I didn't want to say that you should replace `forEach` with `map`. I just described why I think that `forEach` is different than all the other methods (e.g. `map`).

So you should replace `forEach` with a simple loop:

```ts
for (const x of y) {
  ...
}
```